### PR TITLE
[WIP]Use gcs pub sub instead of redis module pub sub

### PR DESCRIPTION
**Pull request by ffbin**
## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 1]
```diff
@@ -549,22 +597,67 @@ Status ServiceBasedNodeInfoAccessor::AsyncSubscribeBatchHeartbeat(
...
-                                                                on_subscribe, done);
+  auto status = gcs_sub_.SubscribeAll(heartbeat_batch_channel_, on_subscribe, done);
   RAY_LOG(DEBUG) << "Finished subscribing batch heartbeat.";
   return status;
 }
 
+void ServiceBasedNodeInfoAccessor::HandleNotification(const GcsNodeInfo &node_info) {
+  ClientID node_id = ClientID::FromBinary(node_info.node_id());
+  bool is_alive = (node_info.state() == GcsNodeInfo::ALIVE);
+  // It's possible to get duplicate notifications from the client table, so
```
Comment (raulchen) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:613]: why would we receive duplicate notifications?

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 2]
```diff
@@ -549,22 +597,67 @@ Status ServiceBasedNodeInfoAccessor::AsyncSubscribeBatchHeartbeat(
...
+      RAY_CHECK(!is_alive)
+          << "Notification for addition of a client that was already removed:" << node_id;
+    }
+  }
+
+  // Add the notification to our cache. Notifications are idempotent.
+  RAY_LOG(INFO) << "[ClientTableNotification] ClientTable Insertion/Deletion "
+                   "notification for client id "
+                << node_id << ". IsAlive: " << is_alive
+                << ". Setting the client cache to data.";
```
Comment (raulchen) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:638]: This seems to be copy-pasted from the old `table.cc` file. We should update this, as we don't have "ClientTable" now.

Comment (raulchen) [reply src/ray/gcs/gcs_client/service_based_accessor.cc:638]: So are other comments that use the term "client".

File: src/ray/protobuf/gcs.proto [hunk 3]
```diff
@@ -311,6 +311,16 @@ message ObjectTableDataList {
   repeated ObjectTableData items = 1;
 }
 
+message ObjectChange {
+  GcsChangeMode change_mode = 1;
```
Comment (raulchen) [inline src/ray/protobuf/gcs.proto:315]: Better just use a boolean "is_add"

File: src/ray/gcs/gcs_server/node_info_handler_impl.h [hunk 4]
```diff
@@ -57,8 +61,32 @@ class DefaultNodeInfoHandler : public rpc::NodeInfoHandler {
...
   gcs::GcsNodeManager &gcs_node_manager_;
+  gcs::GcsPubSub gcs_pub_;
+  const std::string node_channel_ = TablePubsub_Name(TablePubsub::CLIENT_PUBSUB);
+  const std::string node_resource_channel_ =
+      TablePubsub_Name(TablePubsub::NODE_RESOURCE_PUBSUB);
+
+  /// Mutex to protect the nodes_cache_ field.
+  absl::Mutex node_mutex_;
+  /// A mapping from node id to node info.
+  std::unordered_map<ClientID, rpc::GcsNodeInfo> nodes_cache_ GUARDED_BY(node_mutex_);
```
Comment (raulchen) [inline src/ray/gcs/gcs_server/node_info_handler_impl.h:84]: I think we should put this in `GcsNodeManager`.

File: src/ray/gcs/gcs_client/service_based_accessor.h [hunk 5]
```diff
@@ -95,15 +97,12 @@ class ServiceBasedActorInfoAccessor : public ActorInfoAccessor {
...
-  ClientID subscribe_id_;
-
  private:
   ServiceBasedGcsClient *client_impl_;
 
-  typedef SubscriptionExecutor<ActorID, ActorTableData, ActorTable>
-      ActorSubscriptionExecutor;
-  ActorSubscriptionExecutor actor_sub_executor_;
+  /// Subscribe to GCS table information.
+  GcsPubSub gcs_sub_;
```
Comment (raulchen) [inline src/ray/gcs/gcs_client/service_based_accessor.h:104]: I think we can just use one `GcsPubSub` instance (which should be owned by the gcs client) for different accessors.

File: src/ray/gcs/gcs_client/service_based_accessor.h [hunk 6]
```diff
@@ -46,8 +47,9 @@ class ServiceBasedJobInfoAccessor : public JobInfoAccessor {
  private:
   ServiceBasedGcsClient *client_impl_;
 
-  typedef SubscriptionExecutor<JobID, JobTableData, JobTable> JobSubscriptionExecutor;
-  JobSubscriptionExecutor job_sub_executor_;
+  /// Subscribe to GCS table information.
+  GcsPubSub gcs_sub_;
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.h]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.h [hunk 7]
```diff
@@ -46,8 +47,9 @@ class ServiceBasedJobInfoAccessor : public JobInfoAccessor {
...
+  const std::string job_channel_ = TablePubsub_Name(TablePubsub::JOB_PUBSUB);
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.h]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 8]
```diff
@@ -314,17 +320,19 @@ Status ServiceBasedNodeInfoAccessor::UnregisterSelf() {
...
         if (status.ok()) {
           local_node_info_.set_state(GcsNodeInfo::DEAD);
           local_node_id_ = ClientID::Nil();
         }
         RAY_LOG(DEBUG) << "Finished unregistering node info, status = " << status
                        << ", node id = " << node_id;
+        promise.set_value(status);
       });
-  return Status::OK();
+  return promise.get_future().get();
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc:336]: UnregisterSelf  is called when a raylet is destructed, so makes this API synchronous.

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 9]
```diff
@@ -391,33 +399,63 @@ Status ServiceBasedNodeInfoAccessor::AsyncSubscribeToNodeChange(
...
+  };
+
+  // Callback to request notifications from the client table once we've
+  // successfully subscribed.
+  auto on_done = [this, subscribe, done](Status status) {
+    RAY_CHECK_OK(status);
+    if (done != nullptr) {
+      done(status);
+    }
+    // Register node change callbacks after RequestNotification finishes.
```
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 1 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 2 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: fixed

File: src/ray/protobuf/gcs.proto (hunk 3 above)
Comment (ffbin) [reply src/ray/protobuf/gcs.proto]: fixed

File: src/ray/gcs/gcs_server/node_info_handler_impl.h (hunk 4 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/node_info_handler_impl.h]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.h (hunk 5 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.h]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 10]
```diff
@@ -200,7 +209,11 @@ Status ServiceBasedActorInfoAccessor::AsyncSubscribe(
 Status ServiceBasedActorInfoAccessor::AsyncUnsubscribe(const ActorID &actor_id,
                                                        const StatusCallback &done) {
   RAY_LOG(DEBUG) << "Cancelling subscription to an actor, actor id = " << actor_id;
-  auto status = actor_sub_executor_.AsyncUnsubscribe(subscribe_id_, actor_id, done);
+  auto status =
+      client_impl_->GetGcsPubSub().Unsubscribe(ACTOR_CHANNEL, actor_id.Binary());
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:213]: Is this a synchronous operation? Why not invoke the callback of `done` after the `Unsubscribe` is finished synchronous?

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (wumuzi520) [reply src/ray/gcs/gcs_client/service_based_accessor.cc:336]: I am not quite sure does it work. I think the `promise.get_future().get();` will block the main thread, and the `promise.set_value(status);` inside the callback will never be invoked as the callback is also executed in the main thread.

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 11]
```diff
@@ -391,33 +400,50 @@ Status ServiceBasedNodeInfoAccessor::AsyncSubscribeToNodeChange(
     const StatusCallback &done) {
   RAY_LOG(DEBUG) << "Subscribing node change.";
   RAY_CHECK(subscribe != nullptr);
-  ClientTable &client_table = client_impl_->GetRedisGcsClient().client_table();
-  auto status = client_table.SubscribeToNodeChange(subscribe, done);
+  RAY_CHECK(node_change_callback_ == nullptr);
+  node_change_callback_ = subscribe;
+
+  auto on_subscribe = [this, subscribe](const std::string &id, const std::string &data) {
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:406]: There is no need to capture the `subscribe` as it is  not used.

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 12]
```diff
@@ -391,33 +400,50 @@ Status ServiceBasedNodeInfoAccessor::AsyncSubscribeToNodeChange(
...
+    GcsNodeInfo node_info;
+    node_info.ParseFromString(data);
+    HandleNotification(node_info);
+  };
+
+  // Callback to request notifications from the client table once we've
+  // successfully subscribed.
+  auto on_done = [subscribe, done](const Status &status) {
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:414]: There is no need to capture the `subscribe` as it is  not used.

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 13]
```diff
@@ -391,33 +400,50 @@ Status ServiceBasedNodeInfoAccessor::AsyncSubscribeToNodeChange(
...
   return status;
 }
 
 boost::optional<GcsNodeInfo> ServiceBasedNodeInfoAccessor::Get(
     const ClientID &node_id) const {
-  GcsNodeInfo node_info;
-  ClientTable &client_table = client_impl_->GetRedisGcsClient().client_table();
-  bool found = client_table.GetClient(node_id, &node_info);
+  RAY_CHECK(!node_id.IsNil());
+  auto entry = node_cache_.find(node_id);
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:429]: I think this logic could be more simple like:
```
auto entry = node_cache_.find(node_id);
if (entry != node_cache_.end()) {
  return entry->second;
}
return boost::optional::none;
```

BTW, I am not quit  sure if it is ok to get the node info from cache directly.

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 14]
```diff
@@ -628,16 +702,24 @@ Status ServiceBasedTaskInfoAccessor::AsyncSubscribe(
...
+                                                       on_subscribe, done);
   RAY_LOG(DEBUG) << "Finished subscribing task, task id = " << task_id;
   return status;
 }
 
 Status ServiceBasedTaskInfoAccessor::AsyncUnsubscribe(const TaskID &task_id,
                                                       const StatusCallback &done) {
   RAY_LOG(DEBUG) << "Unsubscribing task, task id = " << task_id;
-  auto status = task_sub_executor_.AsyncUnsubscribe(subscribe_id_, task_id, done);
+  auto status = client_impl_->GetGcsPubSub().Unsubscribe(TASK_CHANNEL, task_id.Binary());
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:719]: Is this a synchronous operation? Why not invoke the callback of done after the Unsubscribe is finished synchronous?

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 15]
```diff
@@ -669,16 +751,25 @@ Status ServiceBasedTaskInfoAccessor::AsyncSubscribeTaskLease(
...
   RAY_LOG(DEBUG) << "Finished subscribing task lease, task id = " << task_id;
   return status;
 }
 
 Status ServiceBasedTaskInfoAccessor::AsyncUnsubscribeTaskLease(
     const TaskID &task_id, const StatusCallback &done) {
   RAY_LOG(DEBUG) << "Unsubscribing task lease, task id = " << task_id;
-  auto status = task_lease_sub_executor_.AsyncUnsubscribe(subscribe_id_, task_id, done);
+  auto status =
+      client_impl_->GetGcsPubSub().Unsubscribe(TASK_LEASE_CHANNEL, task_id.Binary());
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:769]: ditto.

File: src/ray/gcs/gcs_client/service_based_accessor.cc [hunk 16]
```diff
@@ -792,16 +881,30 @@ Status ServiceBasedObjectInfoAccessor::AsyncSubscribeToLocations(
...
   RAY_LOG(DEBUG) << "Finished subscribing object location, object id = " << object_id;
   return status;
 }
 
 Status ServiceBasedObjectInfoAccessor::AsyncUnsubscribeToLocations(
     const ObjectID &object_id, const StatusCallback &done) {
   RAY_LOG(DEBUG) << "Unsubscribing object location, object id = " << object_id;
-  auto status = object_sub_executor_.AsyncUnsubscribe(subscribe_id_, object_id, done);
+  auto status =
+      client_impl_->GetGcsPubSub().Unsubscribe(OBJECT_CHANNEL, object_id.Binary());
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_client/service_based_accessor.cc:904]: ditto.

File: src/ray/gcs/gcs_server/gcs_node_manager.cc [hunk 17]
```diff
@@ -15,17 +15,20 @@
...
+#include <ray/gcs/redis_client.h>
 #include <ray/protobuf/gcs.pb.h>
 
 namespace ray {
 namespace gcs {
 GcsNodeManager::GcsNodeManager(boost::asio::io_service &io_service,
                                gcs::NodeInfoAccessor &node_info_accessor,
-                               gcs::ErrorInfoAccessor &error_info_accessor)
+                               gcs::ErrorInfoAccessor &error_info_accessor,
+                               std::shared_ptr<gcs::RedisClient> redis_client)
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/gcs_node_manager.cc:26]: Don't rely on `RedisClient` directly here. You could replace it with `gcs::GcsPubSub`.

File: src/ray/gcs/gcs_server/gcs_server.cc [hunk 18]
```diff
@@ -135,13 +136,13 @@ void GcsServer::InitGcsActorManager() {
 }
 
 std::unique_ptr<rpc::JobInfoHandler> GcsServer::InitJobInfoHandler() {
-  return std::unique_ptr<rpc::DefaultJobInfoHandler>(
-      new rpc::DefaultJobInfoHandler(*redis_gcs_client_));
+  return std::unique_ptr<rpc::DefaultJobInfoHandler>(new rpc::DefaultJobInfoHandler(
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/gcs_server.cc:139]: I think you can add a member of gcs::GcsPubSub to  `GcsServer` and pass it as a parameter to each handler.

File: src/ray/gcs/gcs_server/node_info_handler_impl.h [hunk 19]
```diff
@@ -57,8 +60,25 @@ class DefaultNodeInfoHandler : public rpc::NodeInfoHandler {
...
+                       const SendReplyCallback &send_reply_callback);
+
   gcs::RedisGcsClient &gcs_client_;
   gcs::GcsNodeManager &gcs_node_manager_;
+  gcs::GcsPubSub gcs_pub_sub_;
+
+  /// Mutex to protect the nodes_cache_ field.
+  absl::Mutex node_resource_mutex_;
+  /// A mapping from node id to node resource.
+  std::unordered_map<ClientID, gcs::NodeInfoAccessor::ResourceMap> nodes_resource_cache_
```
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/node_info_handler_impl.h:80]: I think it is better to add a `GcsResourceManager`` to handle the resource related logic.

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: I have remove promise.get_future().get(); thanks

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 10 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc:213]: Unsubscribe not support callback now

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 11 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 12 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 13 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc:419]: fixed

File: src/ray/gcs/gcs_client/service_based_accessor.cc (hunk 14 above)
Comment (ffbin) [reply src/ray/gcs/gcs_client/service_based_accessor.cc:705]: Unsubscribe not support callback now.

File: src/ray/gcs/gcs_server/gcs_node_manager.cc (hunk 17 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_node_manager.cc]: This is a good idea, i have fix it.

File: src/ray/gcs/gcs_server/gcs_server.cc (hunk 18 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_server.cc]: fixed

File: src/ray/gcs/gcs_server/node_info_handler_impl.h (hunk 19 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/node_info_handler_impl.h]: fixed

File: src/ray/gcs/gcs_server/gcs_node_resource_manager.h [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+
+#include <ray/common/id.h>
+#include "ray/gcs/accessor.h"
+
+namespace ray {
+namespace gcs {
+
+/// GcsNodeResourceManager is responsible for managing node resource.
+/// This class is thread-safe.
+class GcsNodeResourceManager {
```
[the remark this answers is not in the document]
Comment (wumuzi520) [reply src/ray/gcs/gcs_server/gcs_node_resource_manager.h:26]: > This class is relatively simple. I think maybe we can just put these methods in GcsNodeManager.

Previously, I thought of an independent GCS resource manager. Later, I found that it was relatively simple. I already put a `cluster_resources_` in GcsNodeManager.
Pls see this PR. https://github.com/ray-project/ray/pull/8119

**Pull request closed without merging**